objects/src/accounts/delta/storage.rs

```diff
@@ -20,224 +26,135 @@ const MAX_MUTABLE_STORAGE_SLOT_IDX: u8 = 254;
-    where
-        A: IntoIterator<Item = u8>,
-        B: IntoIterator<Item = (u8, Word)>,
-        C: IntoIterator<Item = (u8, StorageMapDelta)>,
-    {
+    pub fn from_iters(
+        cleared_items: impl IntoIterator<Item = u8>,
+        updated_items: impl IntoIterator<Item = (u8, Word)>,
+        updated_maps: impl IntoIterator<Item = (u8, StorageMapDelta)>,
+    ) -> Self {
         Self {
-            cleared_items: Vec::from_iter(cleared_items),
-            updated_items: Vec::from_iter(updated_items),
-            updated_maps: Vec::from_iter(updated_maps),
+            slots: BTreeMap::from_iter(
+                cleared_items.into_iter().map(|key| (key, EMPTY_WORD)).chain(updated_items),
+            ),
+            maps: BTreeMap::from_iter(updated_maps),
         }
     }
 
+    /// Returns a reference to the updated slots in this storage delta.
+    pub fn slots(&self) -> &BTreeMap<u8, Word> {
+        &self.slots
+    }
+
+    /// Returns a reference to the updated maps in this storage delta.
+    pub fn maps(&self) -> &BTreeMap<u8, StorageMapDelta> {
+        &self.maps
+    }
+
+    /// Returns true if storage delta contains no updates.
+    pub fn is_empty(&self) -> bool {
+        self.slots.is_empty() && self.maps.is_empty()
+    }
+
     /// Merges another delta into this one, overwriting any existing values.
-    pub fn merge(self, other: Self) -> Result<Self, AccountDeltaError> {
+    pub fn merge(&mut self, other: Self) -> Result<(), AccountDeltaError> {
         self.validate()?;
         other.validate()?;
 
-        let items =
-            self.cleared_items
-                .into_iter()
-                .map(|slot| (slot, None))
-                .chain(self.updated_items.into_iter().map(|(slot, value)| (slot, Some(value))))
-                .chain(other.cleared_items.into_iter().map(|slot| (slot, None)).chain(
-                    other.updated_items.into_iter().map(|(slot, value)| (slot, Some(value))),
-                ))
-                .collect::<BTreeMap<_, _>>();
-
-        let cleared_items = items
-            .iter()
-            .filter_map(|(slot, value)| value.is_none().then_some(*slot))
-            .collect();
-        let updated_items =
-            items.iter().filter_map(|(slot, value)| value.map(|v| (*slot, v))).collect();
+        self.slots.extend(other.slots);
 
-        let mut updated_maps = BTreeMap::<u8, StorageMapDelta>::new();
-        for (slot, update) in self.updated_maps.into_iter().chain(other.updated_maps.into_iter()) {
-            let entry = updated_maps.entry(slot).or_default();
-            *entry = entry.clone().merge(update);
+        // merge maps
+        for (slot, update) in other.maps.into_iter() {
+            match self.maps.entry(slot) {
+                Entry::Vacant(entry) => {
+                    entry.insert(update);
+                },
+                Entry::Occupied(mut entry) => entry.get_mut().merge(update),
+            }
```

Comment:
```suggestion
            self.maps.entry(slot)
              .and_modify(|x| x.merge(update))
              .or_insert(update);
```

Reply:
Even though the proposed change looks better, this wouldn't work because the `update` is moved in the first call (`and_modify`) and it's used after moving in `or_insert`.